src/backend/impl/src/services/access_control_service.rs

```diff
@@ -0,0 +1,207 @@
+use crate::repositories::{UserConfig, UserProfileRepository, UserProfileRepositoryImpl};
+use backend_api::ApiError;
+use candid::Principal;
+
+#[cfg_attr(test, mockall::automock)]
+pub trait AccessControlService {
+    fn assert_principal_not_anonymous(&self, calling_principal: &Principal)
+        -> Result<(), ApiError>;
+
+    async fn assert_principal_is_admin(&self, calling_principal: Principal)
+        -> Result<(), ApiError>;
+}
+
+pub struct AccessControlServiceImpl<T: UserProfileRepository> {
+    user_profile_repository: T,
+}
+
+impl Default for AccessControlServiceImpl<UserProfileRepositoryImpl> {
+    fn default() -> Self {
+        Self::new(UserProfileRepositoryImpl::default())
+    }
+}
+
+impl<T: UserProfileRepository> AccessControlService for AccessControlServiceImpl<T> {
+    fn assert_principal_not_anonymous(
+        &self,
+        calling_principal: &Principal,
+    ) -> Result<(), ApiError> {
+        if calling_principal == &Principal::anonymous() {
+            return Err(ApiError::unauthenticated());
+        }
+
+        Ok(())
+    }
+
+    async fn assert_principal_is_admin(
```

Comment:
This isn't used yet in this PR, but it will be used in a follow-up PR. This is also the reason why I moved `assert_principal_not_anonymous` into a service. Since we need to access the stable structures to assert a principal is an admin, I felt it should be in a service and I also felt that the not-anonymous assertion should be in the same place.